bounds.h

```diff
@@ -0,0 +1,209 @@
+  // constructors when Intervals is empty.
+  template <size_t N = sizeof...(Intervals), class = std::enable_if_t<(N > 0)>>
+  NDARRAY_HOST_DEVICE bounds(const Intervals&... intervals)
+      : intervals_(
+            internal::assert_intervals_compatible<intervals_type>(std::make_tuple(intervals...))) {}
+
+  NDARRAY_HOST_DEVICE bounds(const bounds&) = default;
+  NDARRAY_HOST_DEVICE bounds(bounds&&) = default;
+  NDARRAY_HOST_DEVICE bounds& operator=(const bounds&) = default;
+  NDARRAY_HOST_DEVICE bounds& operator=(bounds&&) = default;
+
+  /** Construct or assign a bounds from another set of intervals of a possibly
+   * different type. Each interval must be compatible with the corresponding
+   * interval of this bounds. */
+  template <class... OtherIntervals, class = enable_if_intervals_compatible<OtherIntervals...>>
+  NDARRAY_HOST_DEVICE bounds(const std::tuple<OtherIntervals...>& other)
+      : intervals_(internal::assert_intervals_compatible<intervals_type>(other)) {}
+  // ELEPHANT - ambiguous with the second constructor?
+  // template <class... OtherIntervals, class = enable_if_intervals_compatible<OtherIntervals...>>
+  // NDARRAY_HOST_DEVICE bounds(OtherIntervals... other_intervals)
+  //  : bounds(std::make_tuple(other_intervals...)) {}
+  template <class... OtherIntervals, class = enable_if_intervals_compatible<OtherIntervals...>>
+  NDARRAY_HOST_DEVICE bounds(const bounds<OtherIntervals...>& other)
+      : intervals_(internal::assert_intervals_compatible<intervals_type>(other.intervals())) {}
+  template <class... OtherIntervals, class = enable_if_intervals_compatible<OtherIntervals...>>
+  NDARRAY_HOST_DEVICE bounds& operator=(const bounds<OtherIntervals...>& other) {
+    intervals_ = internal::assert_intervals_compatible<intervals_type>(other.intervals());
+    return *this;
+  }
+
+  /** Get a specific interval `I` of this bounds. */
+  template <size_t I, class = enable_if_interval<I>>
+  NDARRAY_HOST_DEVICE auto& interval() {
+    return std::get<I>(intervals_);
+  }
+  template <size_t I, class = enable_if_interval<I>>
+  NDARRAY_HOST_DEVICE const auto& interval() const {
+    return std::get<I>(intervals_);
+  }
+
+  /** Get a specific interval of this bounds with a runtime interval index `d`.
+   * This will lose knowledge of any compile-time constant interval
+   * attributes, and it is not a reference to the original interval. */
+  NDARRAY_HOST_DEVICE nda::interval<> interval(size_t i) const {
+    assert(i < rank());
+    return internal::tuple_to_array<nda::interval<>>(intervals_)[i];
+  }
+
+  /** Get a tuple of all of the intervals of this bounds. */
+  NDARRAY_HOST_DEVICE intervals_type& intervals() { return intervals_; }
+  NDARRAY_HOST_DEVICE const intervals_type& intervals() const { return intervals_; }
+
+  /** Makes a tuple of dims out of each interval in this bounds. */
+  NDARRAY_HOST_DEVICE auto dims() const { return internal::dims(intervals(), interval_indices()); }
```

Comment:
`bounds` is just a `shape` without most of the features. It has only min and extents, either of which can be static.

`dims()` is the only fun function, which converts to a `tuple<dim<...>, dim, dim, ...>` appropriately, which can then be used to construct a `shape`.